keps/sig-node/1977-container-notifier/README.md

```diff
@@ -0,0 +1,773 @@
+        Error *ActionError
+}
+
+type ActionError struct {
+        // Type of the error
+        // This is required
+        ErrorType ErrorType
+
+        // Error message
+        // +optional
+        Message *string
+
+        // More detailed reason why error happens
+        // +optional
+        Reason *string
+	
+        // It indicates when the error occurred
+        // +optional
+        Timestamp *int64
+}
+
+type ErrorType string
+
+// More error types could be added, e.g., Forbidden, Unauthorized, etc.
+const (
+        // The Notification times out
+        Timeout ErrorType = "Timeout"
+
+        // The Notification fails with an error
+        Error ErrorType = "Error"
+)
+```
+
+External controller deletes the Notification API objects after it is done.
+
+### Phase 2 API Additions
+
+In Phase 2, add a "signal" field to ContainerNotifierHandler.
+
+```
+// Handler defines a specific action that should be taken or a signal that should be delivered
+// Only one of Exec and Signal should be set and not both
+type ContainerNotifierHandler struct {
+    // Exec specifies the action to take.
+    // +optional
+    Exec *ExecAction `json:"exec,omitempty" protobuf:"bytes,1,opt,name=exec"`
+
+    // Signal specifies a signal to send to the container
+    // +optional
+    // define constants for signals?
+    // validate the signals are valid?  windows?
+++    Signal string
+}
+```
+
+### Phase 3 API Additions
+
+If needed, we may add a Probe to ContainerNotifier in Phase 3, to verify that the ContainerNotifierAction has actually delivered the signal or has run the command.
+
+Phase 3 is more loosely decided than phases 1 and 2, and if it is needed, we will have to revisit the design details.
+
+```
+type ContainerNotifier struct {
+   // Name must be unique within a container.
+   // Names are label-style, with an optional prefix. Names without the prefix
+   // such as `quiesce` are reserved for Kubernetes-defined “well-known” names.
+   // Names with a prefix such as `example.com/label-style` are custom names.
+-   // This refers to a ContainerNotifierAction and a Probe.
++   // This refers to a ContainerNotifierAction and a Probe.
++   // Kubelet executes the command defined in ContainerNotifierAction, and then
++   // calls the command defined in Probe to verify whether the action has  
++   // resulted in a desired state.
++   // For example, if the ContainerNotifierAction has a command to quiesce the application,
++   // then the Probe has a command to verify that the application is indeed
++   // being quiesced. Kubelet will run the command to do a quiesce and wait for the quiescent
++   // probe to return success.
+   Name string
+   Action *ContainerNotifierAction
++   // Add a Probe in Phase 2. Probe defined in the core API here will be used.
++   Probe *Probe
+}
+```
+
+### Kubelet Impact in Phase 2 and Beyond
+
+When moving the logic from the container notifier controller to Kubelet, there will be impact on Kubelet.
```

Comment:
its not clear from this proposal why the controller is starting with a separate controller(also not clear if its separate or part of one of the existing controllers) and then later moving to kubelet

Reply:
This will start as a separate controller in a new repo, and later on the logic will be moved to kubelet.
This is because there are concerns from sig-node on potential impact on kubelet.

Reply:
An abundance of caution